cmp/grid/GridStateModel.js

```diff
@@ -0,0 +1,196 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2018 Extremely Heavy Industries Inc.
+ */
+import {XH, HoistModel} from '@xh/hoist/core';
+import {cloneDeep, debounce, find, uniqBy} from 'lodash';
+import {SECONDS} from '@xh/hoist/utils/DateTimeUtils';
+import {start, resolve} from '@xh/hoist/promise';
+
+@HoistModel()
+export class GridStateModel {
+    parent = null;
+    xhStateId = null;
+
+    state = {};
+    defaultState = null;
+
+    /**
+    * @param {string} xhStateId - Unique grid identifier.
+    * @param {string} [trackColumns] - Save column visible state and ordering.
+    * @param {string} [trackSort] - Save grid sort.
+    */
+    constructor({xhStateId, trackColumns = true, trackSort = true}) {
+        this.xhStateId = xhStateId;
+        this.trackColumns = trackColumns;
+        this.trackSort = trackSort;
+    }
+
+    init(gridModel) {
+        this.parent = gridModel;
+
+        this.ensureCompatible();
+
+        if (this.trackColumns) {
+            this.addReaction(this.columnReaction());
+        }
+
+        if (this.trackSort) {
+            this.addReaction(this.sortReaction());
+        }
+
+        this.initializeState();
+    }
+
+    initializeState() {
+        const userState = this.readState(this.getStateKey());
+        this.defaultState = this.readStateFromGrid();
+
+        this.loadState(userState);
+    }
+
+    readStateFromGrid() {
+        return {
+            columns: this.getColumnState(),
+            sortBy: this.parent.sortBy
+        };
+    }
+
+    //--------------------------
+    // For Extension / Override
+    //--------------------------
+    readState(stateKey) {
+        return XH.localStorageService.get(stateKey, {});
+    }
+
+    saveState(stateKey, state) {
+        XH.localStorageService.set(stateKey, state);
+    }
+
+    resetState(stateKey) {
+        XH.localStorageService.remove(stateKey);
+    }
+
+    loadState(state) {
+        this.state = cloneDeep(state || this.readState(this.getStateKey()) || {});
+        this.updateGridColumns();
+        this.updateGridSort();
+    }
+
+    resetStateAsync() {
+        const defaultState = this.defaultState;
+
+        if (!defaultState) resolve();
+
+        return start(() => {
+            this._resetting = true;
+            this.loadState(defaultState);
+        }).then(() => {
+            this.resetState(this.getStateKey());
+        }).finally(() =>{
+            this._resetting = false;
+        });
+    }
```

Comment:
Not sure if this method needs to be async. It is in Hoist-Sencha, but I think that maybe for a db based override to loading/reseting state. Not sure if we want to prepare for that now or keep it simpler for the moment.